src/tools/crawler/crawler.rs

```diff
@@ -1,107 +1,100 @@
+pub(super) async fn crawl(addr: SocketAddr, known_network: Arc<KnownNetwork>) {
+    tokio::spawn(async move {
+        trace!("Crawling {}", addr);
+        if known_network.new_node(addr).await {
+            match get_crawl_response(addr).await {
+                Ok((response, duration)) => {
+                    process_response(addr, response, duration, known_network).await
                 }
                 Err(e) => {
-                    error!("invalid address from {}: {}", addr, e);
+                    warn!("Unable to get crawl response from {}: {:?}", addr, e);
+                    // TODO if it's connection refused or timeout: retry connection a few time after a while
                 }
             }
         }
-    }
-
-    async fn process_peer(&mut self, addr: SocketAddr, hops: u32, peer: SocketAddr) {
-        self.known_network.insert_node(peer).await;
-        if hops == 0 {
-            self.known_network.insert_connection(addr, peer).await;
-        }
-        self.get_peers(peer).await;
-    }
-
-    pub(super) async fn get_peers(&self, addr: SocketAddr) {
-        let node = self.node.clone();
-        let network = self.known_network.clone();
+    });
+}
 
-        tokio::spawn(async move {
-            if network.insert_node(addr).await {
-                let start = Instant::now();
-                if connect_node(node, addr).await {
-                    network.set_connected(addr, start.elapsed()).await;
-                }
+/// Processes /crawl response: updates `known_network` and starts more crawling recursively.
+fn process_response(
+    from: SocketAddr,
+    response: CrawlResponse,
+    connecting_time: Duration,
+    known_network: Arc<KnownNetwork>,
+) -> BoxFuture<'static, ()> {
+    // Wrapped in BoxFuture to allow for async recursion.
+    async move {
+        known_network
+            .update_stats(from, connecting_time, response.server.build_version)
+            .await;
+        for peer in response.overlay.active {
+            if let Some(addr) = parse_peer_addr(from, peer) {
+                known_network.insert_connection(from, addr).await;
+                crawl(addr, known_network.clone()).await;
             }
-        });
+        }
     }
+    .boxed()
 }
 
-async fn connect_node(node: InnerNode, addr: SocketAddr) -> bool {
-    let connection_result = timeout(CONNECTION_TIMEOUT, node.connect(addr)).await;
-    match connection_result {
-        Ok(response) => match response {
-            Ok(_) => {
-                trace!("connected to {}", addr);
-                true
-            }
-            Err(e) => {
-                warn!("unable to connect to {} due to {}", addr, e);
-                false
+/// Tries to parse address information from response.
+/// On success returns Some(SocketAddr) on failure logs error and returns None.
+fn parse_peer_addr(from: SocketAddr, peer: Peer) -> Option<SocketAddr> {
+    let ip = match peer.ip.as_ref() {
```

Comment:
do we desire such level of error detail? I'd just go with https://doc.rust-lang.org/std/net/enum.SocketAddr.html#impl-FromStr-for-SocketAddr for the entire address at once

Reply:
I could do:
`SocketAddr::from_str(format!("{}:{}", peer.ip.as_ref()?, peer.port().ok()?).as_str()).ok()`
and be done, but then we lose information why it didn't work, so we can't log it.
Is that ok?

Reply:
hence my question about the level of detail we need; for a crawler it is likely that we'll typically have dozens - if not hundreds - of connections at a time, so any individual connection errors are likely to be of little interest; that being said I have no strong feelings here

Reply:
Agreed and done.